ws-client/src/transport.rs

```diff
@@ -174,64 +176,80 @@ impl<'a> WsTransportClientBuilder<'a> {
 			Mode::Plain => None,
 		};
 
-		let mut err = None;
-		for sockaddr in &self.target.sockaddrs {
-			match self.try_connect(*sockaddr, &connector).await {
-				Ok(res) => return Ok(res),
-				Err(e) => {
-					log::debug!("Failed to connect to sockaddr: {:?} with err: {:?}", sockaddr, e);
-					err = Some(Err(e));
-				}
-			}
-		}
-		// NOTE(niklasad1): this is most likely unreachable because [`Url::socket_addrs`] doesn't
-		// return an empty `Vec` if no socket address was found for the host name.
-		err.unwrap_or(Err(WsHandshakeError::NoAddressFound(self.target.host)))
+		self.try_connect(connector).await
 	}
 
 	async fn try_connect(
-		&self,
-		sockaddr: SocketAddr,
-		tls_connector: &Option<crate::tokio::TlsConnector>,
+		self,
+		mut tls_connector: Option<crate::tokio::TlsConnector>,
 	) -> Result<(Sender, Receiver), WsHandshakeError> {
-		// Try establish the TCP connection.
-		let tcp_stream = {
-			let socket = TcpStream::connect(sockaddr);
-			let timeout = crate::tokio::sleep(self.timeout);
-			futures::pin_mut!(socket, timeout);
-			match future::select(socket, timeout).await {
-				future::Either::Left((socket, _)) => {
-					let socket = socket?;
-					if let Err(err) = socket.set_nodelay(true) {
-						log::warn!("set nodelay failed: {:?}", err);
-					}
-					match tls_connector {
-						None => TlsOrPlain::Plain(socket),
-						Some(connector) => {
-							let dns_name = crate::tokio::DNSNameRef::try_from_ascii_str(&self.target.host)?;
-							let tls_stream = connector.connect(dns_name, socket).await?;
-							TlsOrPlain::Tls(tls_stream)
+		let mut sockaddrs = self.target.sockaddrs;
+		let mut path = PathBuf::from(self.target.path);
+		let mut host = self.target.host;
+		let mut host_header = self.target.host_header;
+
+		let mut err = Err(None);
+
+		let client = loop {
+			let sockaddr = match sockaddrs.pop() {
```

Comment:
dq: shouldn't we try all the addresses in `sockaddrs` in case the `connect` call fails?

Reply:
right thanks

Reply:
I was wrong here, basically if pop fails it means that we have tried all sockaddrs.

Reply:
Hmm, when `pop()` succeeds here we proceed to `connect()` (L199), but if that fails we exit the `try_connect()` function (because of `.await?`). Shouldn't we instead try with the next item in `sockaddrs`?

Reply:
yupp, that makes sense fixed.